### PR TITLE
Add Azure AD Auth Provider

### PR DESCRIPTION
**Pull request by ContrarianChris**
## Hey, I just made a Pull Request!

I've managed to get a semi-working implementation going for Azure AD as an OIDC auth provider but have hit some blockers. Not so much with my understanding of how Backstage auth works, though a few things go a bit beyond me, more with how [passport-azure-ad](https://github.com/AzureAD/passport-azure-ad) behaves.

Opening this PR as draft to share what I've done so far and hopefully get some help on resolving what is left and finishing things off. I have added a bunch of comments to the changes to add some more info and get some feedback on how I've gone about things.

### Progress So Far

Azure option added to Sign In Page...

azure-signin-page

Azure AD login prompt...

azure-ad-login-window

Azure AD consent prompt...

azure-ad-consent-window

Header message with Backstage ID, using local part of email...

header-message

User settings showing Backstage profile display name and picture, plus entry for Azure provider...

user-settings

#### Refresh Token Strategy
Hitting refresh attempts to call the `/refresh` endpoint which triggers `executeRefreshTokenStrategy`, which currently doesn't do anything so you are returned to the Sign In Page and need to auth again. 

Looking at the Google implementation it looks like `passport-google-oauth20` uses `passport-oauth2` under the hood which provides the `getAccessToken` function that handles exchanging the refresh token, but that is as far as I got.

There is an open issue for `passport-azure-ad` to support refreshing tokens (AzureAD/passport-azure-ad/issues/297), though it was opened over 3 years ago with no progress as yet. There are a number of threads around the place talking to using [MSAL.js](https://github.com/AzureAD/microsoft-authentication-library-for-js) on top of Passport.js to handle refreshing tokens but I'm not sure how that would work with Backstage.

#### Hijacked State Parameter
Backstage uses the `state` parameter to encode a nonce value and the environment and then verifies the return of that when handling the auth response. `passport-azure-ad` hijacks the `state` parameter and builds its own value which is what is then returned.

"Custom" state can be provided through a `customState` parameter. Doing this...
```
const providerOptions = {
  ...options,
  prompt: 'consent',
  customState: options.state,
};
return await executeRedirectStrategy(req, this._strategy, providerOptions);
```

`passport-azure-ad` then sets the `state` parameter to this...
```
let state = params.state = oauthConfig.customState ? ('CUSTOM' + aadutils.uid(32) + oauthConfig.customState) : aadutils.uid(32);
```

Which gives us a return value of...
```
CUSTOM9K1R2fWvHxxD2KB4RmDKsmhuBNGmw3QHnonce%3Da0Jzscgg7yAx%252FnncHa3CTg%253D%253D%26env%3Ddevelopment
```
Which then causes `readState` in `OAuthProvider` to throw an invalid state error. We effectively need to `.substring(38)` this returned state value to get the original value that Backstage is expecting. 

This is obviously unique to the Azure provider and the others need the `readState` behaviour to remain unchanged. So the question is how best to handle this?

#### Core-API Implementation `defaultScopes`?
What are the `defaultScopes` defined as part of `sessionManager` within `AzureAuth.ts` used for?

I ask as they don't seem to have an affect when working with `passport-azure-ad`, as it needs scopes declared in the `createAzureProvider` function and passed to `new AzureAuthProvider` as part of the options. This seems to be different from the other provider implementations which don't have scopes defined within `provider.ts`.

Wondering whether that is down to my implementation, and/or what fallout there might be that I haven't realised.

Thanks!

#### :heavy_check_mark: Checklist

- [ ] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [x] Relevant documentation updated
- [x] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

File: plugins/auth-backend/src/lib/OAuthProvider.ts [hunk 1]
```diff
@@ -41,8 +41,12 @@ export type Options = {
 };
 
 const readState = (stateString: string): OAuthState => {
+
+  // This is hardcoded for Azure (breaks other providers) until workaround is found
+  const fixedState = stateString.substring(38)
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/lib/OAuthProvider.ts:46]: This is the hardcoded workaround I've put in to get the Azure provider functional.

File: packages/core-api/src/apis/implementations/auth/azure/AzureAuth.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+              Date.now() + res.providerInfo.expiresInSeconds * 1000,
+            ),
+          },
+        };
+      },
+    });
+
+    const sessionManager = new RefreshingAuthSessionManager({
+      connector,
+      defaultScopes: new Set(['openid', 'offline_access', 'profile', 'email']),
```
Comment (ContrarianChris) [inline packages/core-api/src/apis/implementations/auth/azure/AzureAuth.ts:101]: I have set these to match what is defined in the backend `provider.ts` file, but unsure of their use or impact on thing as they don't seem to have an affect on `passport-azure-ad` behaviour.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  makeProfileInfo,
+} from '../../lib/PassportStrategyHelper';
+
+import { OAuthProvider } from '../../lib/OAuthProvider';
+import { Logger } from 'winston';
+import { TokenIssuer } from '../../identity';
+import { Config } from '@backstage/config';
+import passport from 'passport';
+
+import got from 'got';
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:44]: Used to fetch the user's photo, which is not returned in the token and must be retrieved from the Microsoft Graph API's `/me` endpoint.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+      displayName: rawProfile.displayName,
+      provider: 'azure',
+      name: {
+        familyName: rawProfile.name.familyName,
+        givenName: rawProfile.name.givenName,
+      },
+      // If no email for user, fallback to preferred_username
+      emails: [
+        {
+          value: rawProfile._json.email || rawProfile._json.preferred_username,
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:71]: I put this in primarily as most of my test users don't have email's (no Office 365) and I wanted to be able to login. Might make more sense to ultimately leave this undefined in the event of no email address.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  }
+
+  async start(
+    req: express.Request,
+    options: Record<string, string>,
+  ): Promise<RedirectInfo> {
+    const providerOptions = {
+      ...options,
+      prompt: 'consent',
+      customState: options.state,
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:150]: This passing the encoded `state` value into the `customState` parameter of `passport-azure-ad`.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+      OAuthResponse,
+      PrivateInfo
+    >(req, this._strategy);
+    return {
+      response: await this.populateIdentity(response),
+      refreshToken: privateInfo.refreshToken,
+    };
+  }
+
+  async refresh(refreshToken: string, scope: string): Promise<OAuthResponse> {
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:168]: This currently does not work.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+  const responseMode = 'query';
+  const redirectUrl = `${baseUrl}/${providerId}/handler/frame`;
+
+  /**
+   * Space separated list of required scope values. 'openid' is sent by default so does not need to be specified here.
+   * For more info, see:
+   * - https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent
+   * - https://docs.microsoft.com/en-us/graph/permissions-reference
+   */
+  const scope = 'offline_access profile email User.Read';
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:235]: Scopes for `passport-azure-ad` need to be set here.

File: plugins/auth-backend/src/providers/azure/provider.ts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+  const passReqToCallback = true;
+
+  // passport-azure-ad requires use of cookies to be session-free
+  const useCookieInsteadOfSession = true;
+  const cookieEncryptionKeys = [
+    { key: '12345678901234567890123456789012', iv: '123456789012' },
+    { key: 'abcdefghijklmnopqrstuvwxyzabcdef', iv: 'abcdefghijkl' },
```
Comment (ContrarianChris) [inline plugins/auth-backend/src/providers/azure/provider.ts:243]: `passport-azure-ad` requires `express-session` or the use of cookies to be sessionless.

**Comment by javaniecampbell** [comment]
@ContrarianChris Can you let me know how I can get access to the information listed in #1967 please so I can build on what you have for the pipelines plugin I have in place please?

**Comment by ContrarianChris** [comment]
Abandoning this in favour of #2056

**Pull request closed without merging**